keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,784 @@
+The existing way to acceess `finalizers` field through `metadata.finalizers` remains unchanged.
+
+### User Stories (Optional)
+
+#### Story 1
+
+There is a controller that only updates `status` field and `finalizers` field of a CRD. Administrators of the controller can allow the controller to update only these particular fields.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Name collision of the endpoints is needed to be considered.
+- Other existing fields may require the same alternative way to access.
+For example, [candidates](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go) in metadata fields are:
+  - `labels`
+  - `annotations`
+  - `ownerReferences`
+  - `status/conditions`
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Add `/{resource_name}/finalizers` endpoint to all resources and allow access to `metadata.finalizers` for the resources through the endpoints.
+The actual field in each resource isn't added, but the request to the endpoint is handled as the operation to `metadata.finalizers`.
+`Role` and `ClusterRole` can be set to the endpoint, therefore users can set a separate RBAC to `/{resource_name}/finalizers`.
+A PoC implementation can be found [here](https://github.com/mkimuram/kubernetes/commit/751a67e03d62397c1d21947e9e2e3a81f279a29f) (Please also see [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) for how it works. Note that API definition in this KEP will differ from the one in the PoC).
+
+### API Changes
+
+A new `Finalize` resource is introduced in `subresources` API group:
+
+```golang
+// Finalize represents a finalizers field to be accessed via subresource.
+type Finalize struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Only the `finalizers` field is updateable.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+```
+
+### Supported API calls to the endpoint
+
+The `/{resource_name}/finalizers` endpoint supports 3 verbs, GET, PATCH, and PUT.
+
+#### GET
+
+GET verb is used to read the value in `metadata.finalizers` field.
+
+- Request format:
+Empty string is passed as a request.
+
+- Response format:
+  - On success, the `Finalize` resource is returned.
+    ex)
+    ```json
+    {
+        "kind": "Finalize",
+        "apiVersion": "v1alpha1",
+        "metadata": {
```

Comment:
Are we stripping the whole metadata object before serving it? How would `creationTimestamp` be null? Where is the rest of the data? Where is the name?

Reply:
We can choose to populate anything we want here.  We could return the whole ObjectMeta or a subset of fields.

It seems to me like returning *everything* kind of misses the point of access control, but this stuff really ought not be thought of as sensitive.

What about:
* Name
* Namespace
* UID
* RV
* Finalizers

This omits labels, annotations, owners, generation, timestamps, etc.

Name and namespace are redundant but consistent.  UID and RV allow for learning about preconditions.

@apelisse I don't know about ManagedFields?  I just don't have enough context

Reply:
Deleted `creationTimestamp` and added `Namespace`, `Name`, `UID`, and `ResourceVersion`.